### PR TITLE
Add signal handler to close gnmi server

### PR DESCRIPTION
**Pull request by zbud-msft**
#### Why I did it

When process exits or killed unexpectedly, there is a possibility that gnmi server on declared port may still be open. When process received SIGTERM or SIGKILL, we can now stop gnmi server on port.

/var/log$ sudo netstat -nap | grep 50051
tcp6       0      0 :::50051                :::*                    LISTEN      24588/telemetry

Feb  5 08:34:42.363614 INFO telemetry#supervisord: telemetry E0205 08:34:42.358957      19 telemetry.go:176] Failed to create gNMI server: failed to open listener port 50051: listen tcp :50051: bind: address already in use

#### How I did it

Add signal handler via goroutine with channel and when syscalls are made, write to channel to close server.

#### How to verify it

UT

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

### CODE REVIEW

File: gnmi_server/server.go [hunk 1]
```diff
@@ -177,6 +177,15 @@ func (srv *Server) Serve() error {
 	return srv.s.Serve(srv.lis)
 }
 
+func (srv *Server) Stop() {
+	s := srv.s
+	if s == nil {
+		log.Errorf("Stop() failed: not initialized")
+		return
+	}
+	srv.s.Stop()
```
Comment (qiluo-msft) [inline gnmi_server/server.go:186]: >srv.s [](http://example.com/codeflow?start=1&length=5)

Is it just `s`?

**Comment by qiluo-msft** [comment]
How to verify whether "port may still be open"? Do you want to add a sonic-mgmt testcase?

File: telemetry/telemetry.go [hunk 2]
```diff
@@ -183,11 +186,34 @@ func main() {
 	if *withMasterArbitration {
 		s.ReqFromMaster = gnmi.ReqFromMasterEnabledMA
 	}
-	
-	log.V(1).Infof("Auth Modes: ", userAuth)
+
+	var reload = make(chan int, 1)
```
Comment (ganglyu) [inline telemetry/telemetry.go:190]: Why do you add this channel? Add some comments?

File: telemetry/telemetry.go [hunk 3]
```diff
@@ -183,11 +186,34 @@ func main() {
...
+
+	wg.Add(1)
+
+	go signalHandler(reload, &wg, nil)
+
+	value := <-reload
+	log.V(1).Infof("Received Notification for gnmi server to shutdown")
+	s.Stop()
+	if value == 0 {
+		os.Exit(0)
```
Comment (ganglyu) [inline telemetry/telemetry.go:213]: If we exit here, will we lose syslog?

File: telemetry/telemetry.go [hunk 4]
```diff
@@ -209,3 +235,17 @@ func getflag(name string) string {
...
 	return val
 }
+
+func signalHandler(reload chan<- int, wg *sync.WaitGroup, testSigChan <-chan os.Signal) {
+	defer wg.Done()
+	sigchannel := make(chan os.Signal, 1)
+	signal.Notify(sigchannel,
+		syscall.SIGTERM,
+		syscall.SIGQUIT)
+	select {
```
Comment (ganglyu) [inline telemetry/telemetry.go:245]: Can you explain sigchannel and testSigChan?

File: telemetry/telemetry_test.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+
+	go signalHandler(reload, wg, testSigChan)
+
+	testSigChan <- syscall.SIGTERM
+
+	select {
+	case val := <-reload:
+		if val != 0 {
```
Comment (ganglyu) [inline telemetry/telemetry_test.go:28]: Is it possible to get non-zero value here?

File: telemetry/telemetry_test.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import (
+	"os"
+	"testing"
+	"context"
+	"time"
+	"sync"
+	"syscall"
+)
+
+func TestSignalHandler(t *testing.T) {
```
Comment (ganglyu) [inline telemetry/telemetry_test.go:12]: This test can cover SignalHandler, how do we test main?

Comment (zbud-msft) [reply telemetry/telemetry_test.go:19]: Main function now does not contain any real functional logic besides calling setup flags and goroutines

File: telemetry/telemetry_test.go (hunk 5 above)
Comment (zbud-msft) [reply telemetry/telemetry_test.go:35]: As of right now, no, but with cert rotation in the future we will use a non-zero value to indicate that we should not exit goroutine but stop server, rotate certs, and restart server.

File: telemetry/telemetry.go (hunk 4 above)
Comment (zbud-msft) [reply telemetry/telemetry.go]: Removed

File: telemetry/telemetry.go (hunk 3 above)
Comment (zbud-msft) [reply telemetry/telemetry.go]: Removed

File: telemetry/telemetry.go (hunk 2 above)
Comment (zbud-msft) [reply telemetry/telemetry.go]: Added comments, reload channel is used to as a way for gnmi server to stop. serve() is now executed in a goroutine and we use reload channel such that if it gets a value, we will stop server and exit.

File: gnmi_server/server.go (hunk 1 above)
Comment (zbud-msft) [reply gnmi_server/server.go]: Changed to just s.

**Comment by zbud-msft** [comment]
> How to verify whether "port may still be open"? Do you want to add a sonic-mgmt testcase?

Will create a sonic-mgmt testcase

**Comment by zbud-msft** [comment]
Depends on https://github.com/sonic-net/sonic-gnmi/pull/186 for fixing semgrep

**Pull request closed without merging**